### PR TITLE
Include using-decls in function iteration (ROOT-10789)

### PR DESCRIPTION
**Pull request by Axel-Naumann**

### CODE REVIEW

File: core/metacling/src/TCling.cxx [hunk 1]
```diff
@@ -4102,7 +4102,7 @@ TCling::CheckClassInfo(const char *name, Bool_t autoload, Bool_t isClassOrNamesp
                int lineNumber = 0;
                bool success = false;
                std::tie(success, lineNumber) =
-                  ROOT::TMetaUtils::GetTrivialIntegralReturnValue(implLineFunc.GetMethodDecl(), *fInterpreter);
+                  ROOT::TMetaUtils::GetTrivialIntegralReturnValue(implLineFunc.GetAsFunctionDecl(), *fInterpreter);
```
Comment (pcanal) [inline core/metacling/src/TCling.cxx:4105]: Could GetTrivialIntegralReturnValue return a nullptr (since it has a dyn_cast)?

File: core/metacling/src/TCling.cxx [hunk 2]
```diff
@@ -4877,8 +4877,10 @@ void TCling::GetFunctionOverloads(ClassInfo_t *cl, const char *funcname,
       DName = &Ctx.Idents.get(funcname);
    }
 
+   // NotForRedeclaration: we want to find names in inline namespaces etc.
    clang::LookupResult R(S, DName, clang::SourceLocation(),
-                         Sema::LookupOrdinaryName, clang::Sema::ForRedeclaration);
+                         Sema::LookupOrdinaryName, clang::Sema::NotForRedeclaration);
+   R.suppressDiagnostics();
```
Comment (pcanal) [inline core/metacling/src/TCling.cxx:4883]: Could you add a comment why this is needed here?

File: core/metacling/test/TClingTests.cxx [hunk 3]
```diff
@@ -147,18 +151,35 @@ TEST_F(TClingTests, GetClassSharedLibs)
...
+      std::string ret = s;
+      while (!ret.empty() && std::isspace(ret[0]))
+         ret.erase(0, 1);
+      while (!ret.empty() && std::isspace(ret[ret.size() - 1]))
+         ret.erase(ret.size() - 1, 1);
+      return ret;
+   };
+
+   auto split = [](const std::string &s) -> std::vector<std::string> {
+      std::vector<std::string> ret;
```
Comment (pcanal) [inline core/metacling/test/TClingTests.cxx:165]: Out of curiosity why the switch from `llvm::SmallVector<llvm::StringRef, 32>` to ` std::vector<std::string>` ?

File: core/metacling/src/TClingMethodInfo.h [hunk 4]
```diff
@@ -53,35 +53,99 @@ namespace ROOT {
...
+   using Vec_t = llvm::SmallVector<clang::FunctionDecl *,4>;
+   Vec_t fDefDataSpecFuns; // Special functions materialized from DefinitionData.
+   size_t fIDefDataSpecFuns = 0; // Current element in fDefDataSpecFuns.
+
+public:
+   SpecFuncIter() = default;
+   SpecFuncIter(cling::Interpreter *interp, clang::DeclContext *DC,
+                llvm::SmallVectorImpl<clang::Decl*> &&specFuncs);
+
+   bool Valid() const { return fIDefDataSpecFuns < fDefDataSpecFuns.size(); }
```
Comment (pcanal) [inline core/metacling/src/TClingMethodInfo.h:70]:
```suggestion
   bool IsValid() const { return fIDefDataSpecFuns < fDefDataSpecFuns.size(); }
```
everywhere else we use IsValid.

File: core/metacling/src/TClingMethodInfo.h [hunk 5]
```diff
@@ -53,35 +53,99 @@ namespace ROOT {
...
+         return *fSpecFuncIter;
+      return TClingMemberIter::AdvanceUnfiltered();
+   }
+
+public:
+   TClingCXXRecMethIter() = default;
+   TClingCXXRecMethIter(cling::Interpreter *interp, clang::DeclContext *DC,
+                        llvm::SmallVectorImpl<clang::Decl*> &&specFuncs):
+      TClingMemberIter(interp, DC), fSpecFuncIter(interp, DC, std::move(specFuncs)) {
+      Init();
```
Comment (pcanal) [inline core/metacling/src/TClingMethodInfo.h:106]: This may or may not be doing what you want.  If it does, it ought to be explicitly documented.  
Reason: Init uses the virtual function [Is]Valid

File: core/metacling/src/TClingMethodInfo.cxx [hunk 6]
```diff
@@ -484,10 +430,13 @@ long TClingMethodInfo::Property() const
...
    property |= kIsCompiled;
-   const clang::FunctionDecl *fd = GetMethodDecl();
+   const clang::FunctionDecl *fd = GetTargetFunctionDecl();
    if (fd->isConstexpr())
       property |= kIsConstexpr;
-   switch (fd->getAccess()) {
+   // NOTE: this uses `GetDecl()`, to capture the access of the UsingShadowDecl,
+   // which is defined in the derived class and might differ from the access of fd
+   // in the base class.
+   switch (GetDecl()->getAccess()) {
```
Comment (pcanal) [inline core/metacling/src/TClingMethodInfo.cxx:439]: It feels weird to have to get it twice or are the result of GetDecl and GetTargetFunctionDecl "unrelated"?

File: core/metacling/src/TClingMethodInfo.cxx [hunk 7]
```diff
@@ -53,117 +55,257 @@ compiler, not CINT.
...
       // Assemble special functions (or FunctionTemplate-s) that are synthesized from DefinitionData but
       // won't be enumerated as part of decls_begin()/decls_end().
-      for (auto ctor: SemaRef.LookupConstructors(cxxdecl))
-         emplaceSpecFunIfNeeded(ctor);
-      emplaceSpecFunIfNeeded(SemaRef.LookupCopyingAssignment(cxxdecl, /*Quals*/ 0, /*RValueThis*/ false, 0 /*ThisQuals*/));
-      emplaceSpecFunIfNeeded(SemaRef.LookupMovingAssignment(cxxdecl, /*Quals*/ 0, /*RValueThis*/ false, 0 /*ThisQuals*/));
-      emplaceSpecFunIfNeeded(SemaRef.LookupDestructor(cxxdecl));
+      auto Ctors = SemaRef.LookupConstructors(CXXRD);
+      SpecFuncs.append(Ctors.begin(), Ctors.end());
+      SpecFuncs.emplace_back(SemaRef.LookupCopyingAssignment(CXXRD, /*Quals*/ 0, /*RValueThis*/ false, 0 /*ThisQuals*/));
```
Comment (pcanal) [inline core/metacling/src/TClingMethodInfo.cxx:263]: Why was "emplaceSpecFunIfNeeded" needed before and not now?

File: core/metacling/src/TClingMethodInfo.cxx [hunk 8]
```diff
@@ -53,117 +55,257 @@ compiler, not CINT.
...
+         fDefDataSpecFuns.emplace_back(llvm::dyn_cast<clang::FunctionDecl>(D));
+      }
+   };
 
-   return *this;
+   for (auto SpecFunc: specFuncs)
+      emplaceSpecFunIfNeeded(SpecFunc);
 }
 
+bool TClingCXXRecMethIter::Skip(const clang::Decl* D) const
```
Comment (pcanal) [inline core/metacling/src/TClingMethodInfo.cxx:86]: Skip is an action but the function return Yes/No.  
```suggestion
bool TClingCXXRecMethIter::ShouldSkip(const clang::Decl* D) const
```
would result in:
```
   If (ShouldSkip(decl))
     continue;
```

File: core/metacling/src/TCling.cxx (hunk 1 above)
Comment (Axel-Naumann) [reply core/metacling/src/TCling.cxx:4105]: I don't understand your question. I'd understand if it were
> Could GetAsFunctionDecl return a nullptr (since it has a dyn_cast)?

`implLineFunc.IsValid()` is checked before; we do not support using decls for the `ImplLine` function, so by definition it must be a `FunctionDecl`.

File: core/metacling/test/TClingTests.cxx (hunk 3 above)
Comment (Axel-Naumann) [reply core/metacling/test/TClingTests.cxx:165]: To remove the dependency on libCling internals; see the corresponding commit log.

File: core/metacling/src/TCling.cxx (hunk 1 above)
Comment (pcanal) [reply core/metacling/src/TCling.cxx:4105]: fair enough.

File: core/metacling/src/TClingMethodInfo.cxx (hunk 6 above)
Comment (Axel-Naumann) [reply core/metacling/src/TClingMethodInfo.cxx:439]: Yes, and there's a `NOTE` above which explains that? :thinking:

File: core/metacling/src/TClingMethodInfo.cxx (hunk 8 above)
Comment (Axel-Naumann) [reply core/metacling/src/TClingMethodInfo.cxx]: Thanks, fixed.

File: core/metacling/src/TClingMethodInfo.cxx (hunk 6 above)
Comment (pcanal) [reply core/metacling/src/TClingMethodInfo.cxx:450]: Ok.  So they are related but getting from `GetDecl` to `GetTargetFunctionDecl` is non-trivial (conditional loop over using statement) ... and GetDecl is "fast" (even-though it something calls `GetDeclSlow`), so it is okay to call it twice ...

right?

Comment (Axel-Naumann) [reply core/metacling/src/TClingMethodInfo.cxx:450]: Yes:

iter              | `GetDecl()` | `GetTargetFunctionDecl()`
----------------|----------------|-------------------------------
`FunctionDecl` | iter | iter
`UsingShadowDecl` | iter | `iter->getTargetDecl()`

And "Slow" is indeed not *that* slow.

**Comment by Axel-Naumann** [comment]
Failures can be ignored (CDash submission timeout)

**Review by pcanal — APPROVED** [review]

File: core/metacling/test/TClingTests.cxx [hunk 9]
```diff
@@ -93,52 +95,54 @@ TEST_F(TClingTests, MakeInterpreterValue)
...
 
    // FIXME: We should probably resolve again to GenVector as it contains the
    // template pattern.
    lib = GetLibs("ROOT::Math::PxPyPzE4D<int>");
-   EXPECT_STREQ("", MakeLibNamePlatformIndependent(lib).c_str());
+   EXPECT_EQ("", MakeLibNamePlatformIndependent(lib));
 
    lib = GetLibs("vector<ROOT::Math::LorentzVector<ROOT::Math::PxPyPzE4D<double> > >");
-   EXPECT_STREQ("GenVector", MakeLibNamePlatformIndependent(lib).c_str());
+   EXPECT_EQ("GenVector", MakeLibNamePlatformIndependent(lib));
```
Comment (pcanal) [inline core/metacling/test/TClingTests.cxx:137]: Why the change from _STREQ to _EQ?

Comment (pcanal) [reply core/metacling/test/TClingTests.cxx:137]: because the new return type is std::string :)

Comment (Axel-Naumann) [reply core/metacling/test/TClingTests.cxx:137]: It returns a `std::string` anyway?
